Projects/Server/Buffers/STArrayPool.cs

```diff
@@ -172,6 +174,12 @@ private STArray[] InitializeBuckets()
     {
         Debug.Assert(_cacheBuckets is null, $"Non-null {nameof(_cacheBuckets)}");
         var buckets = new STArray[BucketCount];
+
+        if (Interlocked.Exchange(ref _trimCallbackCreated, 1) == 0)
+        {
+            Gen2GcCallback.Register(o => ((STArrayPool<T>)o).Trim(), this);
```

Comment:
We could probably optimize this later by unregistering it if the stacks are empty. It isn't a big concern though since it is a single object per generic type.